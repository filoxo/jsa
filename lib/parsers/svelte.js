const { promises: fs } = require("node:fs");
const { parse } = require("svelte/compiler");

async function parseWithSvelte(file) {
  const fileBlob = await fs.readFile(file, { encoding: "utf-8" });
  const { instance: ast } = parse(fileBlob);
  return ast;
}

module.exports = {
  parseWithSvelte,
};
