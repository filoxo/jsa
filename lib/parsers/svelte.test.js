import { describe, expect, it, vi } from "vitest";
import { parseWithSvelte } from "./svelte";
import { promises as fs } from "node:fs";

describe("ast-worker", () => {
  const spyReadFile = vi.spyOn(fs, "readFile");

  it("returns an ast when parsing plain javascript", async () => {
    spyReadFile.mockResolvedValue(
      `<script>
import { onDestroy, onMount } from 'svelte';
export let name
</script>

<style>
section {
  font-size: 1.5rem;
}
</style>

<section>{name} is mounted!</section>`
    );

    const result = await parseWithSvelte("/path/to/App.svelte");
    expect(result).toMatchSnapshot();
  });
});
