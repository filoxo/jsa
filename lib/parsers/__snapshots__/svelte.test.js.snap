// Vitest Snapshot v1

exports[`ast-worker > returns an ast when parsing plain javascript 1`] = `
{
  "content": Node {
    "body": [
      Node {
        "end": 53,
        "loc": SourceLocation {
          "end": Position {
            "column": 44,
            "line": 2,
          },
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "source": Node {
          "end": 52,
          "loc": SourceLocation {
            "end": Position {
              "column": 43,
              "line": 2,
            },
            "start": Position {
              "column": 35,
              "line": 2,
            },
          },
          "raw": "'svelte'",
          "start": 44,
          "type": "Literal",
          "value": "svelte",
        },
        "specifiers": [
          Node {
            "end": 27,
            "imported": Node {
              "end": 27,
              "loc": SourceLocation {
                "end": Position {
                  "column": 18,
                  "line": 2,
                },
                "start": Position {
                  "column": 9,
                  "line": 2,
                },
              },
              "name": "onDestroy",
              "start": 18,
              "type": "Identifier",
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 18,
                "line": 2,
              },
              "start": Position {
                "column": 9,
                "line": 2,
              },
            },
            "local": Node {
              "end": 27,
              "loc": SourceLocation {
                "end": Position {
                  "column": 18,
                  "line": 2,
                },
                "start": Position {
                  "column": 9,
                  "line": 2,
                },
              },
              "name": "onDestroy",
              "start": 18,
              "type": "Identifier",
            },
            "start": 18,
            "type": "ImportSpecifier",
          },
          Node {
            "end": 36,
            "imported": Node {
              "end": 36,
              "loc": SourceLocation {
                "end": Position {
                  "column": 27,
                  "line": 2,
                },
                "start": Position {
                  "column": 20,
                  "line": 2,
                },
              },
              "name": "onMount",
              "start": 29,
              "type": "Identifier",
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 27,
                "line": 2,
              },
              "start": Position {
                "column": 20,
                "line": 2,
              },
            },
            "local": Node {
              "end": 36,
              "loc": SourceLocation {
                "end": Position {
                  "column": 27,
                  "line": 2,
                },
                "start": Position {
                  "column": 20,
                  "line": 2,
                },
              },
              "name": "onMount",
              "start": 29,
              "type": "Identifier",
            },
            "start": 29,
            "type": "ImportSpecifier",
          },
        ],
        "start": 9,
        "type": "ImportDeclaration",
      },
      Node {
        "declaration": Node {
          "declarations": [
            Node {
              "end": 69,
              "id": Node {
                "end": 69,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 15,
                    "line": 3,
                  },
                  "start": Position {
                    "column": 11,
                    "line": 3,
                  },
                },
                "name": "name",
                "start": 65,
                "type": "Identifier",
              },
              "init": null,
              "loc": SourceLocation {
                "end": Position {
                  "column": 15,
                  "line": 3,
                },
                "start": Position {
                  "column": 11,
                  "line": 3,
                },
              },
              "start": 65,
              "type": "VariableDeclarator",
            },
          ],
          "end": 69,
          "kind": "let",
          "loc": SourceLocation {
            "end": Position {
              "column": 15,
              "line": 3,
            },
            "start": Position {
              "column": 7,
              "line": 3,
            },
          },
          "start": 61,
          "type": "VariableDeclaration",
        },
        "end": 69,
        "loc": SourceLocation {
          "end": Position {
            "column": 15,
            "line": 3,
          },
          "start": Position {
            "column": 0,
            "line": 3,
          },
        },
        "source": null,
        "specifiers": [],
        "start": 54,
        "type": "ExportNamedDeclaration",
      },
    ],
    "end": 70,
    "loc": SourceLocation {
      "end": Position {
        "column": 0,
        "line": 4,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 8,
    "type": "Program",
  },
  "context": "default",
  "end": 79,
  "start": 0,
  "type": "Script",
}
`;
